File: src/main/java/com/giffardtechnologies/restdocs/LinkTool.java
package com.giffardtechnologies.restdocs;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.giffardtechnologies.restdocs.domain.NamedEnumeration;
import org.apache.commons.text.StringEscapeUtils;

import com.giffardtechnologies.restdocs.domain.DataObject;
import com.giffardtechnologies.restdocs.domain.Document;

public class LinkTool {
	
	private final Document mDocument;
	private final Set<String> mDataObjectNames;
	
	public LinkTool(Document document) {
		super();
		mDocument = document;
		
		ArrayList<DataObject> dataObjects = mDocument.getDataObjects();
		ArrayList<NamedEnumeration> enumerations = mDocument.getEnumerations();
		mDataObjectNames = new HashSet<String>((dataObjects.size() + enumerations.size()) * 2);
		for (DataObject dataObject : dataObjects) {
			mDataObjectNames.add(dataObject.getName());
		}
		for (NamedEnumeration enumeration : enumerations) {
			mDataObjectNames.add(enumeration.getName());
		}
		if (mDocument.getService() != null && mDocument.getService().hasCommon()) {
			for (DataObject dataObject : mDocument.getService().getCommon().getResponseDataObjects()) {
				mDataObjectNames.add(dataObject.getName());
			}
		}
	}
	
	public String type(String string) {
		if (string == null) {
			return null;
		}
		final String escapeHtml = StringEscapeUtils.escapeHtml4(string);
		Pattern typeLinkPattern = Pattern.compile("&lt;.+?&gt;");
		Matcher matcher = typeLinkPattern.matcher(escapeHtml);
		StringBuffer builder = new StringBuffer();
		while (matcher.find()) {
			final String typeLink = matcher.group();
			String typeName = typeLink.substring(4, typeLink.length() - 4);
			if (mDataObjectNames.contains(typeName)) {
				matcher.appendReplacement(builder, "<a href=\"#" + typeName + "\">" + typeName + "</a>");
			} else {
				matcher.appendReplacement(builder, typeLink);
			}
		}
		matcher.appendTail(builder);
		return builder.toString();
	}

	public String typeSimple(String typeName) {
		if (typeName == null) {
			return null;
		}
		if (mDataObjectNames.contains(typeName)) {
			return "<a href=\"#" + typeName + "\">" + typeName + "</a>";
		} else {
			return typeName;
		}
	}

}

File: src/main/java/com/giffardtechnologies/restdocs/LinkTool.kt
package com.giffardtechnologies.restdocs

import com.giffardtechnologies.restdocs.domain.Document
import org.apache.commons.text.StringEscapeUtils
import java.util.regex.Pattern

class LinkTool(private val mDocument: Document) {
    private val mDataObjectNames: MutableSet<String>

    init {
        val dataObjects = mDocument.dataObjects
        val enumerations = mDocument.enumerations
        mDataObjectNames = HashSet((dataObjects.size + enumerations.size) * 2)
        for (dataObject in dataObjects) {
            mDataObjectNames.add(dataObject.name)
        }
        for (enumeration in enumerations) {
            mDataObjectNames.add(enumeration.name)
        }
        if (mDocument.service != null && mDocument.service.hasCommon()) {
            for (dataObject in mDocument.service.common.responseDataObjects) {
                mDataObjectNames.add(dataObject.name)
            }
        }
    }

    fun type(string: String?): String? {
        if (string == null) {
            return null
        }
        val escapeHtml = StringEscapeUtils.escapeHtml4(string)
        val typeLinkPattern = Pattern.compile("&lt;.+?&gt;")
        val matcher = typeLinkPattern.matcher(escapeHtml)
        val builder = StringBuffer()
        while (matcher.find()) {
            val typeLink = matcher.group()
            val typeName = typeLink.substring(4, typeLink.length - 4)
            if (mDataObjectNames.contains(typeName)) {
                matcher.appendReplacement(builder, "<a href=\"#$typeName\">$typeName</a>")
            } else {
                matcher.appendReplacement(builder, typeLink)
            }
        }
        matcher.appendTail(builder)
        return builder.toString()
    }

    fun typeSimple(typeName: String?): String? {
        if (typeName == null) {
            return null
        }
        return if (mDataObjectNames.contains(typeName)) {
            "<a href=\"#$typeName\">$typeName</a>"
        } else {
            typeName
        }
    }
}
